sylt-compiler/src/name_resolution.rs

```diff
@@ -0,0 +1,1038 @@
+            if var_name == name {
+                return Ok(*var_id);
+            }
+        }
+        match self.lookup_global(namespace_id, name) {
+            Some(Name::Name(var)) => Ok(*var),
+            Some(Name::Namespace(..)) => Err(vec![resolution_error!(
+                self,
+                span,
+                "When resolving the name {:?} - a namespace was found",
+                name
+            )]),
+            None => Err(vec![resolution_error!(
+                self,
+                span,
+                "Failed to resolve {:?} - nothing matched",
+                name
+            )]),
+        }
+    }
+
+    fn namespace_list(&self, namespace_id: usize, assignable: &ParserAssignable) -> Option<usize> {
+        use sylt_parser::AssignableKind as AK;
+        (match &assignable.kind {
+            AK::Read(ident) => match self.lookup_global(namespace_id, &ident.name) {
+                Some(Name::Namespace(file_or_lib, _)) => Some(file_or_lib),
+                _ => None,
+            },
+            AK::Access(prev, ident) => {
+                match self
+                    .namespace_list(namespace_id, prev)
+                    .map(|new_namespace| self.lookup_global(new_namespace, &ident.name))
+                    .flatten()
+                {
+                    Some(Name::Namespace(file_or_lib, _)) => Some(file_or_lib),
+                    _ => None,
+                }
+            }
+            AK::Index { .. }
+            | AK::Variant { .. }
+            | AK::Call { .. }
+            | AK::ArrowCall { .. }
+            | AK::Expression { .. } => None,
+        })
+        .map(|file_or_lib| self.file_to_namespace.get(file_or_lib).cloned())
+        .flatten()
+    }
+
+    fn type_vec(&self, parser_tys: &[ParserType]) -> ResolveResult<Vec<Type>> {
+        let mut tys = Vec::new();
+        let mut errs = Vec::new();
+        for ty in parser_tys.iter() {
+            match self.ty(ty) {
+                Err(mut es) => errs.append(&mut es),
+                Ok(ty) => tys.push(ty),
+            }
+        }
+        if errs.is_empty() {
+            Ok(tys)
+        } else {
+            Err(errs)
+        }
+    }
+
+    fn ty(&self, ty: &ParserType) -> ResolveResult<Type> {
+        use sylt_parser::TypeKind as TK;
+        use Type as T;
+        let span = ty.span;
+        Ok(match &ty.kind {
+            TK::Implied => T::Implied(span),
+            TK::Resolved(t) => T::Resolved(t.clone(), span),
+            TK::UserDefined(t, gs) => {
+                let r = match self.ty_assignable(t)? {
+                    T::UserType(r, _) => r,
+                    _ => raise_resolution_error! {
+                        self,
+                        span,
+                        "This is not a reference to a user defined type"
+                    },
+                };
+                let gs = self.type_vec(gs)?;
+                T::UserDefinedWithGenerics(r, gs, span)
```

Comment:
So if I'm understanding this correctly, `UserDefinedWithGenerics` is used even if `gs` is empty? Then what is `UserType` for? Is there a need for both?

Reply:
In some places you don't want to specify the generics (or maybe you do, but we don't allow it right now). The prime usecase is when instantiating a blob. We write something along the lines of:
`A.B.Blob { ... }`. I'll merge them. Nice catch!